zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/Learner.java

```diff
@@ -792,10 +800,24 @@ boolean isRunning() {
     }
 
     void closeSocket() {
-        try {
-            if (sock != null) {
-                sock.close();
+        if (sock != null) {
+            if (sockBeingClosed.compareAndSet(false, true)) {
+                if (closeSocketAsync) {
+                    final Thread closingThread = new Thread(() -> closeSockSync(), "CloseSocketThread(sid:" + zk.getServerId());
+                    closingThread.setDaemon(true);
+                    closingThread.start();
+                } else {
+                    closeSockSync();
+                }
             }
+        }
+    }
+
+    void closeSockSync() {
+        try {
+            long startTime = Time.currentElapsedTime();
+            sock.close();
```

Comment:
You might want to add 2 more things here:
- double check if sock is still not `null`
- make sock = null after the close to prevent further usage of the object and let GC collect it

Reply:
done.